<!DOCTYPE html>
<html>
  <head>
    <title>Networks and Logoplots</title>
    <meta charset="utf-8">
    <meta name="author" content="Di Cook (dicook@monash.edu, @visnut)   Heike Hofmann (heike.hofmann@gmail.com, @heike_hh)" />
    <link href="index_files/remark-css-0.0.1/example.css" rel="stylesheet" />
    <link rel="stylesheet" href="myremark.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Networks and Logoplots
## SISBID 2017 <br> <a href="https://github.com/SISBID/Module2" class="uri">https://github.com/SISBID/Module2</a>
### Di Cook (<a href="mailto:dicook@monash.edu">dicook@monash.edu</a>, <span class="citation">@visnut</span>) <br> Heike Hofmann (<a href="mailto:heike.hofmann@gmail.com">heike.hofmann@gmail.com</a>, <span class="citation">@heike_hh</span>)
### 07/12-14/2017

---





# General networks

- Parent-child relationships are special forms of networks
- General networks consist of __nodes__ (set of actors or entities) and __edges__ (connections between actors)
- There are lots of packages to draw networks (`igraph`, `sna`, ... )
- Several implementations in `ggplot2` framework (`ggnetwork`, `geomnet` &lt;- the one I'll be using)


---
# Networks

- networks consist of a edges data set and a nodes (vertices) set
- we merge node information into the edges


```r
library(geomnet)
data(blood)

head(blood$edges, 3)
  from  to group_to
1  AB- AB+     same
2  AB- AB-     same
3  AB+ AB+     same

head(blood$vertices, 3)
  label type rho        Ethnicity Predominance
1    A-    A neg            Asian          0.5
2    A-    A neg         Hispanic          2.0
3    A-    A neg African.American          2.0

bloodnet &lt;- left_join(blood$edges, blood$vertices, by=c("from" = "label"))
```

---
# Directed network of blood donation


```r
p &lt;- ggplot(data = bloodnet, aes(from_id = from, to_id = to))
p + geom_net(labelon = TRUE, size = 15, labelcolour="white", 
             vjust=0.5, hjust=0.5, directed=TRUE, arrowgap=0.075) + 
  theme_net()
```

&lt;img src="index_files/figure-html/unnamed-chunk-3-1.png" style="display: block; margin: auto;" /&gt;

---
# Using the ggplot2 framework

- It is possible to adjust node size for another variable, but we are really bad at these kind of comparisons!


```r
p + geom_net(labelon = TRUE, labelcolour="white", 
             vjust=0.5, hjust=0.5, directed=TRUE, arrowgap=0.075,
             aes(size = Predominance)) + theme_net() +
  facet_wrap(~Ethnicity) + scale_size(range=c(5,15)) + 
  expand_limits(x = c(-.1,1.1), y = c(-.1,1.1))
```

&lt;img src="index_files/figure-html/unnamed-chunk-4-1.png" style="display: block; margin: auto;" /&gt;

---
# Bipartite networks

- two types of nodes, vertices are only between nodes of different types

```r
# access the data and rename it for convenience
library(tnet)
data(tnet)
elist &lt;- data.frame(Davis.Southern.women.2mode)
names(elist) &lt;- c("Lady", "Event")

elist$Lady &lt;- paste("L", elist$Lady, sep="")
elist$Event &lt;- paste("E", elist$Event, sep="")

davis &lt;- elist
names(davis) &lt;- c("from", "to")
davis &lt;- rbind(davis, data.frame(from=davis$to, to=davis$from))
davis$type &lt;- factor(c(rep("Lady", nrow(elist)), rep("Event", nrow(elist))))
```


---
# 


```r
# Southern women network in geomnet
davis$lcolour &lt;- 
  c("white", "black")[as.numeric(davis$type)]

ggplot(data = davis) + 
  geom_net(
    aes(from_id = from, to_id = to, 
        colour = type, shape = type), 
    size = 15, labelon = TRUE, ealpha = 0.25,
    vjust = 0.5, hjust = 0.5,
    labelcolour = davis$lcolour) +
  theme_net() + 
  scale_colour_brewer("Type of node", palette = "Set2") +
  scale_shape("Type of node") +
  theme(legend.position = "bottom")
```

---
# Southern women network in geomnet

&lt;img src="index_files/figure-html/unnamed-chunk-8-1.png" style="display: block; margin: auto;" /&gt;

---
# Sequence Logo plots

![](http://weblogo.berkeley.edu/img/cap_hth.png)

- Logo plots (Schneider &amp; Stephens, 1990) show consensus sequences of aligned DNA or peptide sequences
- various (online) apps to create sequence plots, e.g. [Weblogo 3.0](http://weblogo.berkeley.edu/logo.cgi)
- bit dubious: several design aspects  of the plot does not reflect best practice from cognitive perspective

Schneider TD, Stephens RM. Sequence logos: a new way to display consensus sequences. Nucleic Acids Research. 1990;18(20):6097-6100.

---
# Sequence Logos

- sequences are often distributed in the [FASTA file format](https://en.wikipedia.org/wiki/FASTA_format)
- in R `read.fasta` or `read.alignment(format="fasta")` of the `seqinr` package


```r
library(seqinr)
pfam &lt;- read.fasta(file = "../data/pfam.fasta", as.string=TRUE, seqtype="AA", seqonly=TRUE)
head(pfam, 3)
[[1]]
[1] "LTMT.-RGDIGNYLGLTVETISRLLGRFQKLGVL"

[[2]]
[1] "LTMT.-RGDIGNYLGLTVETISR-----------"

[[3]]
[1] "LTMT.-RGDIGNYLGLTVETISR-----------"
```

---
# Construction of Logo plots

In a traditional logo sequence plot:

- x-axis: position of an element,
- y-axis: amount of preserved information (usually Shannon information, in bits)

Shannon information  is an entropy measure.
Formally, information `\(I\)`  in position `\(p\)` is defined as &lt;br&gt;
`\(I(p) = \log_2(k) + \sum_{aa} f_{aa} \log_2(f_{aa}),\)` &lt;br&gt;
where `\(f_{aa}\)` is the relative frequency of amino acid `\(aa\)` in  position `\(p\)` and `\(k\)` is the number of different elements in the sequence ( `\(k = 4\)` for RNA/DNA, and `\(k =21\)` for peptide sequences).

For `\(k = 21\)` (amino-acids): 
`\(I(p)\)` is between 0 and `\(-\log_2 (1/21) = 4.39\)` bits, for perfect diversity (0 bits) and single element only (4.39 bits)

---
# Converting sequences to ggplot2 format

- `ggfortify` processes sequences to a data frame usable in sequence logos in ggplot2 (i.e. introduces `position`, `element`, `info`, ...)
- for peptide sequences includes information such as polarity and water properties


```r
library(gglogo)
ggpfam &lt;- ggfortify(
  data=data.frame(sequences = unlist(pfam)),
  seq="sequences") %&gt;% filter(between(position, 7, 26))

head(ggpfam[order(ggpfam$position),], 3)
   element position freq total       info       bits  Polarity       Water
15       A        7    1   101 0.02469362 0.02469362 non-polar hydrophobic
45       H        7   16   101 0.39509787 0.39509787     basic hydrophobic
50       I        7    1   101 0.02469362 0.02469362 non-polar hydrophobic
```

---


```r
ggplot(data=ggpfam) +
  geom_logo(aes(x = position, label=element, y = bits, fill=interaction(Water, Polarity))) +
  scale_fill_brewer(palette="Paired") +
  theme_bw() +
  theme(legend.position= "bottom")
```

&lt;img src="index_files/figure-html/unnamed-chunk-11-1.png" style="display: block; margin: auto;" /&gt;


---
# GeomLogo vs Weblogo

Cognitively, the approach in GeomLogo is better because ...

- each letter is enveloped within a rectangle (signal independent of the shape of the letter - compare e.g. `W` and `I`)
- rectangles are colored rather than the letters
- the main sequences is emphasized by being the only one along the top
- the secondary sequence is aligned at the top, making it easy to compare as well (Cleveland &amp; McGill, 1984)
- colour scheme and legend

---
# Positioning of rectangles

- default positioning is `position = "logo"`
- `position = "classic"` stacks the rectangles on top of each other, ordered from largest to smallest. This comes closest to the traditional sequence logo plot
- other positions, such as 'stack', 'fill', or 'dodge' work, but might produce strange output.

&lt;img src="index_files/figure-html/unnamed-chunk-12-1.png" style="display: block; margin: auto;" /&gt;



---
# Another example

- `sequences` data set in gglogo package has additional information on treatment/type: gram positive vs gram negative
- `ggfortify` has parameters  `treatment` and `weight`


```r
data(sequences, package = "gglogo")
head(sequences, 3)
                               peptide  class
1 RWTHLASGRTYNYKFNPPKQYGKDDITGEDLIQRED gram -
2 RWTHLNSGRTYHYKFNPPKVHGVDDVTGEPLVQRED gram -
3 RWTHLASGRTYNYKFNPPKQYGKDDITGEDLIQRED gram -
ggseq &lt;- ggfortify(data= sequences, seq = "peptide", treatment = .(class))
head(ggseq, 3)
  element  class position freq total        info        bits Polarity
1       . gram -        1    3   923 0.014173689 0.014173689     &lt;NA&gt;
2       . gram +       21   10   886 0.021908523 0.021908523     &lt;NA&gt;
3       . gram -        2    3   923 0.009847968 0.009847968     &lt;NA&gt;
  Water
1  &lt;NA&gt;
2  &lt;NA&gt;
3  &lt;NA&gt;
```

---
# Facetted Logo plots


```r
ggplot(data = ggseq) +
  geom_logo(aes(x = class, y = info, label = element, fill=interaction(Polarity, Water)), alpha = 0.8) +
  scale_fill_brewer(palette = "Paired") + facet_wrap(~position, ncol=18) +
  theme_bw() +
  theme(legend.position = "bottom")
```

&lt;img src="index_files/figure-html/unnamed-chunk-14-1.png" style="display: block; margin: auto;" /&gt;



---
class: inverse middle 
# Your turn

![](lorikeets.png)

- Create a logo plot for the `sequences` data 
- Use facetting to construct logo plots that come closest to the 2nd (gram +) and 3rd (gram -) logo sequence  shown below

&lt;img width="300" src="http://biovis.net/sites/default/files/adk.png"&gt;

---
# Other parameters

- `ggfortify` has parameter `method` - by default, Shannon information is computed, but `method="freq"` computes frequencies
- parameter `alphabet` in `geom_logo` allows to specify polygons from a different font (output from `createPolygons`)

&lt;img src="index_files/figure-html/unnamed-chunk-15-1.png" style="display: block; margin: auto;" /&gt;

---
# Resources

- [geomnet github repo](https://github.com/sctyner/geomnet)

---
# Share and share alike

This work is licensed under the Creative Commons Attribution-Noncommercial 3.0 United States License. To view a copy of this license, visit http://creativecommons.org/licenses/by-nc/3.0/us/ or send a letter to Creative Commons, 171 Second Street, Suite 300, San Francisco, California, 94105, USA.
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {window.dispatchEvent(new Event('resize'));});</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
